compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs

```diff
@@ -1125,11 +1123,59 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 self.tcx.hir().span_with_body(self.tcx.local_def_id_to_hir_id(fn_id)),
             )
         {
+            // When the expr is in an `arm node`'s body, we shouldn't add semicolon ';' at the end.
+            // For example:
+            // fn mismatch_types() -> i32 {
+            //     match 1 {
+            //         x => dbg!(x),
+            //     }
+            //     todo!()
+            // }
+            // -------------^^^^^^^-
+            // Not add semicolon ';' at the end of expr 'dbg!(x)'
+            fn is_in_arm<'tcx>(expr: &'tcx hir::Expr<'tcx>, tcx: TyCtxt<'tcx>) -> bool {
+                for (_, node) in tcx.hir().parent_iter(expr.hir_id) {
+                    match node {
+                        hir::Node::Block(block) => {
+                            if let Some(ret) = block.expr
+                                && ret.hir_id == expr.hir_id
+                            {
+                                continue;
+                            }
+                        }
+                        hir::Node::Arm(arm) => {
+                            if let hir::ExprKind::Block(block, _) = arm.body.kind
+                                && let Some(ret) = block.expr
+                                && ret.hir_id == expr.hir_id
+                            {
+                                return true;
+                            }
+                        }
+                        hir::Node::Expr(e) if let hir::ExprKind::Block(block, _) = e.kind => {
+                            if let Some(ret) = block.expr
+                                && ret.hir_id == expr.hir_id
+                            {
+                                continue;
+                            }
+                        }
+                        _ => {
+                            return false;
+                        }
+                    }
+                }
+
+                false
+            }
+
             err.multipart_suggestion(
                 "you might have meant to return this value",
                 vec![
                     (span.shrink_to_lo(), "return ".to_string()),
-                    (span.shrink_to_hi(), ";".to_string()),
+                    if !is_in_arm(expr, self.tcx) {
+                        (span.shrink_to_hi(), ";".to_string())
+                    } else {
+                        (span.shrink_to_hi(), "".to_string())
```

Comment:
You can't add an empty suggestion. I would pull this `vec![]` into a separate, and only `suggestions.push()` if the `if` statement is true.

Reply:
> You can't add an empty suggestion. I would pull this `vec![]` into a separate, and only `suggestions.push()` if the `if` statement is true.

Done. Thank you.